### DISCUSSION TITLE
Can we use custom AMP url in link tag instead of "?amp=1"?

### DISCUSSION THREAD

**Discussion by iam1337-dev** [vercel/next.js#38094] [Help]
If we turned on AMP functionality with next/amp, It will generate a link tag with URL appends by ?amp=1 by default. I didn't find any way to change the default amp parameter. I want to generate link tag with URL append by **"?amp"** and not ?amp=1.

Is there any way to do the same?

Generated by default:
![image](https://user-images.githubusercontent.com/92446119/176101427-d2482c23-6eb2-4d24-94d2-ce9b7aa773c6.png)

Requirement:
![image](https://user-images.githubusercontent.com/92446119/176101688-0592b09a-cec6-4f80-a968-e766425664b3.png)

Thanks in advance!

**Comment by icyJoseph** [comment]
Some time, not long ago I helped someone to allow the application to be reachable at `?amp`, only for them to find out that it didn't matter because the `link` used `?amp=1` anyway.

As far as I remember this there is no way to switch to this behaviour. 

I am not an AMP expert, so idk, if there's a reason proper to do `=1`, in the query. The spec on URL queries suggests, `key=value`, pairs, but doesn't make them obligatory, so `?amp` is a valid query, which most parsers evaluate into the `key=value` pair: `amp: ''`

Deep in the code for AMP in Next.js this check exists, https://github.com/vercel/next.js/blob/ccd3df2f812c98874273de7b237baa9c2ad8f309/packages/next/export/worker.ts#L282, which would do `Boolean("")` and that's false so, the way Next.js, at the moment, understands this, `?amp` is no amp mode.

I guess, this is also why the generated links use `?amp=1` - so that it can be properly feedback into the application.

**Reply by iam1337-dev** [reply]
Yes, thanks for the explanation. 
I want **"?amp"** only because my current WordPress website doing the same right now. Now I'm migrating the website into Next.js and end up with conflict with this URL.

If I used ?amp=1 then crawlers will cache 2 amp URLs for 1 record. I thought I will cause an issue with SEO.

**Reply by icyJoseph** [reply]
> I want "?amp" only because my current WordPress website doing the same right now. Now I'm migrating the website into Next.js and end up with conflict with this URL.

You could try with the middleware rewrite trick I talked about before. 

```ts
// _middleware.ts
import { NextResponse, type NextRequest } from "next/server";

export function middleware(req: NextRequest) {
  const amp = req.nextUrl.searchParams.get("amp");
  // strictly check that amp is empty string
  if (amp === "") {
    req.nextUrl.searchParams.set("amp", "1");
    return NextResponse.rewrite(req.nextUrl);
  }

  return NextResponse.next();
}
```

Make sure your pages export amp hybrid, `export const config = { amp: 'hybrid' }`

That way, at least, requests coming to `?amp` still see the Next.js `?amp=1` version, thought this might be unwanted.
 
> If I used ?amp=1 then crawlers will cache 2 amp URLs for 1 record. I thought I will cause an issue with SEO.

I can't speak to that.
